barretenberg/cpp/src/barretenberg/vm/avm/trace/bytecode_trace.cpp

```diff
@@ -26,41 +66,89 @@ std::vector<FF> AvmBytecodeTraceBuilder::pack_bytecode(const std::vector<uint8_t
+            bytecode_hash_trace.push_back(BytecodeHashTraceEntry{
+                .field_encoded_bytecode = field_encoded_bytecode[i],
                 .running_hash = running_hash,
-                .bytecode_length_remaining = static_cast<uint16_t>(packed_bytecode.size() - i),
+                .bytecode_length_remaining = static_cast<uint16_t>(field_encoded_bytecode.size() - i),
             });
-            running_hash = poseidon2::hash({ packed_bytecode[i], running_hash });
+            running_hash = poseidon2::hash({ field_encoded_bytecode[i], running_hash });
         }
         // Now running_hash actually contains the bytecode hash
-        BytecodeTraceEntry last_entry;
-        last_entry.running_hash = running_hash;
+        BytecodeHashTraceEntry last_entry;
         last_entry.bytecode_length_remaining = 0;
-        // TODO: Come back to this later
-        // last_entry.class_id = _
-        // last_entry.contract_address = _
+        last_entry.running_hash = running_hash;
+        ASSERT(running_hash == contract_bytecode.contract_class_id_preimage.public_bytecode_commitment);
+
+        last_entry.class_id = compute_contract_class_id(contract_bytecode.contract_class_id_preimage.artifact_hash,
+                                                        contract_bytecode.contract_class_id_preimage.private_fn_root,
+                                                        running_hash);
+        ASSERT(last_entry.class_id == contract_bytecode.contract_instance.contract_class_id);
+
+        last_entry.contract_address = compute_address_from_instance(contract_bytecode.contract_instance);
+        ASSERT(last_entry.contract_address == contract_bytecode.contract_instance.address);
     }
 }
 
 void AvmBytecodeTraceBuilder::finalize(std::vector<AvmFullRow<FF>>& main_trace)
 {
-    for (size_t i = 0; i < bytecode_trace.size(); i++) {
-        auto const& src = bytecode_trace.at(i);
+    for (size_t i = 0; i < bytecode_hash_trace.size(); i++) {
+        auto const& src = bytecode_hash_trace.at(i);
         auto& dest = main_trace.at(i);
         dest.bytecode_running_hash = src.running_hash;
         dest.bytecode_length_remaining = src.bytecode_length_remaining;
-        dest.bytecode_packed = src.packed_bytecode;
+        dest.bytecode_as_fields = src.field_encoded_bytecode;
         dest.bytecode_end_latch = src.bytecode_length_remaining == 0 ? FF::one() : FF::zero();
     }
+
+    // We should probably combine this step with the previous one
+    size_t row_index = 0;
+    for (auto& bytecode : all_contracts_bytecode) {
+        // Reset the byte index for each contract
+        uint32_t byte_index = 0;
+        for (auto& byte : bytecode.bytecode) {
+            main_trace.at(row_index).bytecode_bytes = byte;
+            main_trace.at(row_index).bytecode_bytes_pc = byte_index;
+        }
+        // The row index will match up with the clk
+        row_index += bytecode.bytecode.size();
+    }
```

Comment:
If I understood `build_bytecode_hash_columns()`, we are computing hashes a full field at a time. So why do we need to create this `bytecode_bytes` column of bytes here?

And since the inner loop isn't incrementing `row_index`, isn't the inner loop just overwriting the same main trace entry over and over?

Reply:
`bytecode_bytes` is the raw byte column that we will eventually use for bytecode decomposition. I thought since we had the info now i might as well put it in the main trace.

And you're right about inner loop! I'll change that